x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/upgrade_prebuilt_rules_table_context.tsx

```diff
@@ -161,16 +193,86 @@ export const UpgradePrebuiltRulesTableContextProvider = ({
     }
   }, [rules, upgradeAllRulesRequest]);
 
+  // Wrapper around upgrade rules methods to display ML Jobs warning modal when necessary
+  const upgradeRulesWrapper = useCallback(
+    (upgradeMethod: ModalConfirmationUpgradeMethod, upgradeRuleMethod: () => Promise<void>) =>
+      () => {
+        if (legacyJobsInstalled.length > 0) {
+          showUpgradeModal();
+          setModalConfirmationUpgradeMethod(upgradeMethod);
+        } else {
+          upgradeRuleMethod();
+        }
+      },
+    [legacyJobsInstalled.length, showUpgradeModal]
+  );
+
+  const upgradeSelectedRulesCTAClick = upgradeRulesWrapper('SPECIFIC_RULES', upgradeSelectedRules);
+  const upgradeAllRulesCTAClick = upgradeRulesWrapper('ALL_RULES', upgradeAllRules);
+
+  // Wrapper around upgradeOneRule to display ML Jobs warning modal when necessary
+  // when attempting to upgrade a single rule from each rule row CTA
+  const upgradeSingleRuleFromRowCTA = useCallback(
+    (ruleId: string) => {
+      if (legacyJobsInstalled.length > 0) {
+        showUpgradeModal();
+        setModalConfirmationUpgradeMethod('SINGLE_RULE');
+        setRuleIdToUpgrade(ruleId);
+      } else {
+        upgradeOneRule(ruleId);
+      }
+    },
+    [
+      legacyJobsInstalled,
+      setModalConfirmationUpgradeMethod,
+      setRuleIdToUpgrade,
+      showUpgradeModal,
+      upgradeOneRule,
+    ]
+  );
+
+  const mlJobUpgradeModalConfirm = useCallback(() => {
+    if (modalConfirmationUpdateMethod === 'ALL_RULES') {
+      upgradeAllRules();
+    } else if (modalConfirmationUpdateMethod === 'SPECIFIC_RULES') {
+      upgradeSelectedRules();
+    } else if (modalConfirmationUpdateMethod === 'SINGLE_RULE') {
+      upgradeOneRule(ruleIdToUpgrade);
+    }
+    hideUpgradeModal();
+  }, [
+    hideUpgradeModal,
+    modalConfirmationUpdateMethod,
+    ruleIdToUpgrade,
+    upgradeAllRules,
+    upgradeOneRule,
+    upgradeSelectedRules,
+  ]);
+
+  const mlJobUpgradeModalCancel = useCallback(() => {
+    setModalConfirmationUpgradeMethod(null);
+    hideUpgradeModal();
+  }, [hideUpgradeModal, setModalConfirmationUpgradeMethod]);
+
   const actions = useMemo<UpgradePrebuiltRulesTableActions>(
     () => ({
       reFetchRules: refetch,
-      upgradeOneRule,
-      upgradeSelectedRules,
-      upgradeAllRules,
+      upgradeSingleRuleFromRowCTA,
+      upgradeSelectedRulesCTAClick,
+      upgradeAllRulesCTAClick,
```

Comment:
It would be ideal to keep the method names neutral and avoid referencing specific use cases. For example, we should remove terms like `Row`, `Click`, or `CTA` from the method names.